tools/base/libboard/board_vc/board_vc.c

```diff
@@ -0,0 +1,749 @@
+			}
+		}
+
+		res = fpgaObjectRead(dev_obj[i], (uint8_t *)cdevid, 0, obj_size, 0);
+		if (res != FPGA_OK) {
+			OPAE_ERR("Failed to read device node");
+			resval = res;
+		}
+
+
+		res = fpgaDestroyObject(&dev_obj[i]);
+		if (res != FPGA_OK) {
+			resval = res;
+			OPAE_ERR("Failed to destroy object");
+		}
+
+		// append null char
+		cdevid[obj_size - 1] = '\0';
+
+		snprintf_s_s(path, sizeof(path), "/dev/char/%s", cdevid);
+
+		fd = open(path, O_RDWR);
+		if (fd < 0) {
+			OPAE_ERR("Open %s failed\n", path);
+			continue;
+		}
+
+		memset_s(&info, sizeof(info), 0);
+		info.argsz = 14;
+
+		if (0 == ioctl(fd, FPGA_PHY_GROUP_GET_INFO, &info)) {
+			group_info[i] = info;
+		} else {
+			OPAE_ERR("ioctl error\n");
+		}
+
+		close(fd);
+
+	} // end for loop
+
+out_destroy_group_dev:
+	res = fpgaDestroyObject(&group_dev_object);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to destroy object");
+	}
+
+out_destroy_group:
+	res = fpgaDestroyObject(&group_object);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to destroy object");
+	}
+
+	return resval;
+
+}
+
+// Read mac information
+fpga_result read_mac_info(fpga_token token, unsigned char *mac_info, size_t len)
+{
+	fpga_result res                 = FPGA_OK;
+	fpga_result resval              = FPGA_OK;
+	unsigned char buf[8]            = {0};
+	fpga_object mac_object;
+
+	if (mac_info == NULL) {
+		FPGA_ERR("Invalid Input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	res = fpgaTokenGetObject(token, SYSFS_EEPROM, &mac_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get token object");
+		return res;
+	}
+
+	res = fpgaObjectRead(mac_object, (uint8_t *)buf, 0, sizeof(buf), 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to read object ");
+		resval = res;
+		goto out_destroy_mac;
+	}
+
+	memcpy_s(mac_info, len, buf, sizeof(buf));
+
+out_destroy_mac:
+	res = fpgaDestroyObject(&mac_object);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to destroy object");
+	}
+
+	return resval;
+}
+
+// Read pkvl versoin
+fpga_result print_pkvl_version(fpga_token token)
+{
+	fpga_result res                 = FPGA_OK;
+	fpga_result resval              = FPGA_OK;
+	char ver_a_buf[VER_BUF_SIZE]    = { 0 };
+	char ver_b_buf[VER_BUF_SIZE]    = { 0 };
+	uint32_t size                   = 0;
+	fpga_object pkvl_a_object;
+	fpga_object pkvl_b_object;
+
+
+	res = fpgaTokenGetObject(token, SYSFS_PKVL_A_VER, &pkvl_a_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_MSG("Failed to get token object");
+		return res;
+	}
+
+	res = fpgaTokenGetObject(token, SYSFS_PKVL_B_VER, &pkvl_b_object, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_MSG("Failed to get token object");
+		resval = res;
+		goto out_destroy_obj_a;
+	}
+
+
+	res = fpgaObjectGetSize(pkvl_a_object, &size, FPGA_OBJECT_GLOB);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get object size");
+		resval = res;
+		goto out_destroy_obj_b;
+	}
+
+	res = fpgaObjectRead(pkvl_a_object, (uint8_t *)ver_a_buf, 0, size, 0);
```

Comment:
Theoretically, the max size of sysfs "files" are usually one page (4096 bytes). 
What if we get some size bigger than 16 when we cal fpgaObjectGetSize on line 538? We will overrun our ver_a_buf.

Reply:
Fixed